File: 20-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c == "(" or c == "[" or c == "{":
                stack.append(c)
                continue

            if len(stack) == 0:
                return False

            top = stack.pop()
            if top == '(' and c != ')' or \
                top == '{' and c != '}' or \
                top == '[' and c != ']':
                return False

        return len(stack) == 0

File: 20-Valid-Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for (char c : s) {
            if (c == '[' || c == '{' || c == '(')
                stk.push(c);
            else {
                if (stk.empty())
                    return false;

                char top = stk.top();
                stk.pop();

                if ((top == '[' && c != ']') || 
                    (top == '{' && c != '}') ||
                    (top == '(' && c != ')'))
                    return false;
            }
        }

        return stk.empty();
    }
};